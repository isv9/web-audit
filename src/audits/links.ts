import { AuditResult, AuditResultLog, WebDocument } from '../web-audit';
import { getEmptyElementsLiveCollections } from './utils';

export function auditLinks(
  document: Pick<WebDocument, 'getEmptyElementsByTagName' | 'getElementsByTagName'>,
): AuditResult {
  const linksElements = document.getElementsByTagName('a');
  const logs: AuditResultLog[] = [];
  if (linksElements.length > 0) {
    logs.push([`Document has ${linksElements.length} links`, linksElements]);
  }
  const warnings: string[] = [];

  const emptyElementsLiveCollections = getEmptyElementsLiveCollections(document, ['a']);
  if (emptyElementsLiveCollections.length > 0) {
    warnings.push('Document has empty links');
  }

  return {
    logs,
    warnings,
    liveCollections: emptyElementsLiveCollections,
    name: 'links in dom',
    tables: [],
  };
}
