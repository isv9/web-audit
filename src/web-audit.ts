import { auditBlockSemantics, auditCommonSemantics, auditHeaderSemantics, auditTextSemantics } from './audits/semantic'
import { auditLinks } from './audits/links'

export type WebDocument = {
  getElementsByTagNameCount (tag: string): number;
  querySelectorAll (query: string): { length: number };
  getEmptyElementsByTagName (tag: string): { length: number }
};

type AuditSectionResult = {
  name: string;
  auditResults: AuditResult[];
};

export type AuditResult = {
  name?: string;
  tables: (AuditResultTable)[];
  liveCollections?: (AuditResultLiveCollection)[];
  logs?: string[];
  warnings?: string[];
  errors?: string[];
};

type AuditResultTable = { name?: string } & { [key: string]: string | number };
export type AuditResultLiveCollection = { name: string, collection: object };

export class WebAudit {

  private readonly webDocument: WebDocument

  constructor (webDocument: WebDocument) {
    this.webDocument = webDocument
  }

  audit () {
    const auditSemanticsResult = this.auditSemantics()
    const auditLinksResult = this.auditLinks()
    console.group('web audit')
    WebAudit.renderAuditSectionResult(auditSemanticsResult)
    WebAudit.renderAuditSectionResult(auditLinksResult)
    console.groupEnd()
  }

  private auditLinks (): AuditSectionResult {
    return {
      name: 'links',
      auditResults: [
        auditLinks(this.webDocument)]
    }
  }

  private auditSemantics (): AuditSectionResult {

    return {
      name: 'semantics',
      auditResults: [auditCommonSemantics(this.webDocument),
        auditHeaderSemantics(this.webDocument),
        auditTextSemantics(this.webDocument),
        auditBlockSemantics(this.webDocument)]
    }
  }

  static renderAuditSectionResult (auditSectionResult: AuditSectionResult) {
    const { name, auditResults } = auditSectionResult
    const details = {
      errorsCount: 0,
      warningsCount: 0
    }
    auditResults.forEach(auditResult => {
      const errorsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'errors')
      const warningsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'warnings')
      details.errorsCount += errorsCount
      details.warningsCount += warningsCount
    })
    console.group(`${name}, (errors="${details.errorsCount}", warnings="${details.warningsCount}")`)
    auditResults.forEach(auditResult => WebAudit.renderAuditResult(auditResult))
    console.groupEnd()
  }

  static renderAuditResult (auditResult: AuditResult) {
    const { name, tables, errors = [], warnings = [], liveCollections = [], logs = [] } = auditResult
    const errorsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'errors')
    const warningsCount = WebAudit.getAuditResultMessagesCount(auditResult, 'warnings')
    console.group(`${name}, (errors="${errorsCount}", warnings="${warningsCount}")`)
    tables.forEach(({ name, nodes, ...tableDataSet }) => {
      console.log(name)
      console.table(tableDataSet)
    })
    liveCollections.forEach(({ name, collection }) => {
      console.log(name)
      console.log(collection)
    })
    logs.forEach(message => console.log(message))
    console.log('Summary')
    if (errorsCount > 0 || warningsCount > 0) {
      errors.forEach(message => console.error(message))
      warnings.forEach(message => console.warn(message))
    } else {
      console.log('It is OK')
    }

    console.groupEnd()
  }

  static getAuditResultMessagesCount (auditResult: AuditResult, key: keyof AuditResult): number {
    let property = auditResult[key]
    if (property === null || property === undefined) {
      property = []
    }
    if (Array.isArray(property)) {
      return property.length
    }
    throw new Error('Property type must be array')
  }
}

