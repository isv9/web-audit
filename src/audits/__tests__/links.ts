import { WebDocument } from '../../web-audit';
import { auditLinks } from '../links';

describe('auditLinks', () => {
  it('document is ok', () => {
    const document: Pick<WebDocument, 'getEmptyElementsByTagName' | 'getElementsByTagName'> = {
      getElementsByTagName: jest.fn(() => ({ length: 3 })),
      getEmptyElementsByTagName: jest.fn(() => ({ length: 0 })),
    };
    expect(auditLinks(document)).toMatchSnapshot();
  });
  it('document has some empty links', () => {
    const document: Pick<WebDocument, 'getEmptyElementsByTagName' | 'getElementsByTagName'> = {
      getElementsByTagName: jest.fn(() => ({ length: 3 })),
      getEmptyElementsByTagName: jest.fn(() => ({ length: 5 })),
    };
    expect(auditLinks(document)).toMatchSnapshot();
  });
});
