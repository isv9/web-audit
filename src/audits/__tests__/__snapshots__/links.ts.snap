// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auditLinks document has some empty links 1`] = `
Object {
  "liveCollections": Array [
    Object {
      "collection": Object {
        "length": 5,
      },
      "name": "Empty \\"a\\" elements count in dom",
    },
  ],
  "logs": Array [
    Array [
      "Document has 3 links",
      Object {
        "length": 3,
      },
    ],
  ],
  "name": "links in dom",
  "tables": Array [],
  "warnings": Array [
    "Document has empty links",
  ],
}
`;

exports[`auditLinks document is ok 1`] = `
Object {
  "liveCollections": Array [],
  "logs": Array [
    Array [
      "Document has 3 links",
      Object {
        "length": 3,
      },
    ],
  ],
  "name": "links in dom",
  "tables": Array [],
  "warnings": Array [],
}
`;
